core/domain/exp_domain.py

```diff
@@ -2829,111 +2951,41 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
-                            for rule_spec in group['rule_specs']:
-                                if is_valid_algebraic_expression(
-                                        rule_spec['inputs']['x']):
-                                    new_rule_specs.append(rule_spec)
-                            group['rule_specs'] = new_rule_specs
-                    else:
-                        new_interaction_id = TYPE_VALID_NUMERIC_EXPRESSION
-
-                    # Removing answer groups that have no rule specs left after
-                    # the filtration done above.
-                    new_answer_groups = [
-                        answer_group for answer_group in new_answer_groups if (
-                            len(answer_group['rule_specs']) != 0)]
-
-                    # Removing feedback keys, from voiceovers_mapping and
-                    # translations_mapping, that correspond to the rules that
-                    # got deleted.
-                    old_answer_groups_feedback_keys = [
-                        answer_group['outcome'][
-                            'feedback']['content_id'] for answer_group in (
-                                state_dict['interaction']['answer_groups'])]
-                    new_answer_groups_feedback_keys = [
-                        answer_group['outcome'][
-                            'feedback']['content_id'] for answer_group in (
-                                new_answer_groups)]
-                    content_ids_to_delete = set(
-                        old_answer_groups_feedback_keys) - set(
-                            new_answer_groups_feedback_keys)
-                    for content_id in content_ids_to_delete:
-                        if content_id in state_dict['recorded_voiceovers'][
-                                'voiceovers_mapping']:
-                            del state_dict['recorded_voiceovers'][
-                                'voiceovers_mapping'][content_id]
-                        if content_id in state_dict['written_translations'][
-                                'translations_mapping']:
-                            del state_dict['written_translations'][
-                                'translations_mapping'][content_id]
+                    for rule_spec in group['rule_specs']:
+                        rule_input = rule_spec['inputs']['x']
+
+                        # Removing all greek letters after adding the
+                        # corresponding symbol to the 'variables' set so that
+                        # the letters in them don't get considered individually.
+                        # For eg. if the expression is 'beta + x', we want the
+                        # variables to be ['β', 'x'] instead of
+                        # ['β', 'b', 'e', 't', 'a', 'x'].
+                        for ind, greek_letter in enumerate(greek_letters):
+                            if greek_letter in rule_input:
+                                rule_input = rule_input.replace(
```

Comment:
The logic seems too ad hoc: does it handle stuff like `betalphaa` correctly? Can you prove that it is equivalent to your parser output?

It would be better to parse this using your parser and then identify the variables etc. from there.

Reply:
Done. I've added a function in the parser file for extracting variables.